### PR TITLE
EVG-13717: add option to apply scopes on enqueue

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/EVG-13717

* Add job option to make a job acquire scopes when it's enqueued (i.e. when `Put()` is called) rather than when the job is dispatched.
* Implement the option for the Mongo driver, which backs the remote queue implementation.

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -84,6 +84,11 @@ type Job interface {
 	// exclusion a job can provide.
 	Scopes() []string
 	SetScopes([]string)
+
+	// ApplyScopesOnEnqueue allows the scope exclusion functionality to be
+	// configured so that exclusion occurs during job dispatch or enqueue.
+	ApplyScopesOnEnqueue() bool
```
Comment (julianedwards) [inline interface.go:90]: This is a bit confusing cause `apply` here can also sound like a verb. I almost want to change this to `IsApplyScopesOnEnqueueSet` or `IsApplyScopesOnEqueue` but maybe that is too long. What do you think?

File: queue/driver_test.go [hunk 2]
```diff
@@ -102,6 +104,91 @@ func (s *DriverSuite) TestPutJobDoesNotAllowDuplicateIds() {
...
+	j = job.NewShellJob("echo bar", "")
+	j.SetScopes([]string{"scope"})
+	j.SetApplyScopesOnEnqueue(true)
+
+	s.Require().NoError(s.driver.Put(s.ctx, j))
+
+	s.Equal(2, s.driver.Stats(s.ctx).Total)
+}
+
+func (s *DriverSuite) TestPutAndSaveJobSucceedsIfScopeIsAppliedOnEnqueue() {
```
Comment (julianedwards) [inline queue/driver_test.go:175]: What is the difference between this test and `TestPutJobAllowsSameScopesInQueueIfInitialScopedJobDoesNotApplyScopeOnEnqueue`?

File: queue/remote_test.go [hunk 3]
```diff
@@ -69,14 +61,15 @@ func TestRemoteUnorderedMongoSuite(t *testing.T) {
 // above Test function and replacing the driverConstructor function.
 
 func (s *RemoteUnorderedSuite) SetupSuite() {
-	s.require = s.Require()
+
```
Comment (julianedwards) [inline queue/remote_test.go:64]: Can we remove this function?

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: queue/driver_test.go (hunk 2 above)
Comment (Kimchelly) [reply queue/driver_test.go:175]: Sorry, the other test was a copy-pasta mistake. I updated it to test the actual `Put()`/`Save()` sequence.

File: queue/remote_test.go (hunk 3 above)
Comment (Kimchelly) [reply queue/remote_test.go]: Removed.

File: interface.go (hunk 1 above)
Comment (julianedwards) [reply interface.go:90]: I agree, `Should` makes the most sense.

Comment (Kimchelly) [reply interface.go]: Done.

**Review by julianedwards — APPROVED** [review]

**Pull request merged**